File: python-algorithm/algo_09_math/L9.py
# -*- coding: utf-8 -*-

# 为了防止有些语言的溢出，这里只反转一半数字
class Solution:
    def isPalindrome(self, x: int) -> bool:
        if x<0: return False
        m = x
        cnt = 0
        # 求出x有多少位
        while m:
            m = m//10
            cnt += 1
        n = x
        k = 0
        # 反转整数的一半
        for _ in range(cnt//2):
            k = k*10 + n%10
            n //= 10
        if cnt&1:
            return n//10 == k
        else:
            return n==k

File: python-algorithm/algo_09_math/L9.cpp

#include <iostream>
using  namespace std;

class Solution {
public:
    bool isPalindrome(int x) {
        if (x < 0)
            return false;
        int rev = 0 , r = x ;
        while (r != 0){
            if (rev > INT_MAX/10 || (rev == INT_MAX/10 && rev%10 > 7 ))
                return false;
            rev = rev*10 + r%10;
            r /= 10;
        }
        return rev == x ? true:false ;
    }
};

//还有一种方法是转化成字符串，从两端往中间比较，或者反转字符串比较。python操作字符串可能更方便一些